### PR TITLE
[SD-1535] Add the name event to the Webhook serializer payload/body

### PR DESCRIPTION
**Pull request by sebastian-palma**
Replaces https://github.com/spree/spree/pull/11469

### CODE REVIEW

**Comment by sebastian-palma** [comment]
Does [this](https://github.com/spree/spree/pull/11469#discussion_r747728491) comment still apply @damianlegawiec?

File: core/lib/spree/testing_support/common_rake.rb [hunk 1]
```diff
@@ -36,6 +36,7 @@
     ]
 
     puts 'Setting up dummy database...'
+    system('bin/rails db:environment:set RAILS_ENV=test')
```
Comment (sebastian-palma) [inline core/lib/spree/testing_support/common_rake.rb:39]: The `test_app` raises an error asking to set the db environment right after some migration fails.

File: api/app/models/concerns/spree/webhooks/has_webhooks.rb [hunk 2]
```diff
@@ -9,20 +9,34 @@ module HasWebhooks
...
         end
       end
 
       private
 
-      def event_name(operation)
-        "#{self.class.name.demodulize.tableize.singularize}.#{operation}"
+      def webhooks_body_for(event:)
+        created_event = event_record(event)
+        created_event.reload
```
Comment (aleksandarpetrushev) [inline api/app/models/concerns/spree/webhooks/has_webhooks.rb:23]: why is this line necessary?

Comment (sebastian-palma) [reply api/app/models/concerns/spree/webhooks/has_webhooks.rb:23]: I think that case is only necessary for the test suite @aleksandarpetrushev. When the event is created in line 22, its timestamps don't coincide with the ones in the database when checking it with AR:

```ruby
Spree::Webhooks::Event.where(name: event)
=> [#<Spree::Webhooks::Event:0x000055a95c90dac0
  id: 19,
  execution_time: nil,
  name: "variant.discontinued",
  request_errors: nil,
  response_code: nil,
  subscriber_id: nil,
  success: nil,
  url: nil,
  created_at: Fri, 19 Nov 2021 15:23:08.683798000 UTC +00:00,
  updated_at: Fri, 19 Nov 2021 15:23:08.683798000 UTC +00:00,
  preferences: nil>]
```

while the `created_event` variable holds this:

```ruby
created_event
=> #<Spree::Webhooks::Event:0x000055a95c8779d0
 id: 19,
 execution_time: nil,
 name: "variant.discontinued",
 request_errors: nil,
 response_code: nil,
 subscriber_id: nil,
 success: nil,
 url: nil,
 created_at: Fri, 19 Nov 2021 15:23:08.683798067 UTC +00:00,
 updated_at: Fri, 19 Nov 2021 15:23:08.683798067 UTC +00:00,
 preferences: {}>
```

as you can see, nor the `created_at` nor the `updated_at` match, but the id does, meaning they should refer to the same record. Just after a `reload` you get that:

```ruby
created_event.reload
=> #<Spree::Webhooks::Event:0x000055a95c8779d0
 id: 19,
 execution_time: nil,
 name: "variant.discontinued",
 request_errors: nil,
 response_code: nil,
 subscriber_id: nil,
 success: nil,
 url: nil,
 created_at: Fri, 19 Nov 2021 15:23:08.683798000 UTC +00:00,
 updated_at: Fri, 19 Nov 2021 15:23:08.683798000 UTC +00:00,
 preferences: nil>
```

Comment (aleksandarpetrushev) [reply api/app/models/concerns/spree/webhooks/has_webhooks.rb:23]: okay 👍

Comment (sebastian-palma) [reply api/app/models/concerns/spree/webhooks/has_webhooks.rb:23]: By the way, this happens only with Postgres.

File: api/app/models/spree/webhooks/event.rb [hunk 3]
```diff
@@ -1,9 +1,10 @@
 module Spree
   module Webhooks
     class Event < Spree::Webhooks::Base
-      validates :name, :subscriber, presence: true
+      validates :name, presence: true
+      validates :subscriber, presence: true, on: :update
 
-      belongs_to :subscriber, inverse_of: :events, optional: false
+      belongs_to :subscriber, inverse_of: :events, optional: true
```
Comment (damianlegawiec) [inline api/app/models/spree/webhooks/event.rb:7]: @sebastian-palma why are we changing this to optional? we're aware of the subscriber before sending the event

Comment (sebastian-palma) [reply api/app/models/spree/webhooks/event.rb:7]: That's right, I just added at the "beginning" of the flow, I'll move it to HandleRequest so the insert goes in the background as well.

File: api/spec/services/spree/webhooks/subscribers/handle_request_spec.rb [hunk 4]
```diff
@@ -39,15 +34,42 @@
...
     end
 
+    before do
+      expect(Spree::Webhooks::Event).to(
+        receive(:create).
+          with(name: event_name, subscriber_id: subscriber.id, url: url).
+          and_return(event)
+      )
+      expect(Spree::Webhooks::Subscribers::MakeRequest).to(
+        receive(:new).with(body: body.to_json, url: url).and_return(make_request_double)
```
Comment (sebastian-palma) [inline api/spec/services/spree/webhooks/subscribers/handle_request_spec.rb:60]: @damianlegawiec this is the only place where the body with the added event data is tested now. [Here](https://github.com/spree/spree/pull/11504/files#diff-726ef97c3571e54a204da8dfb105bf2ee55b61239453fc5b434311eb479eda9cR7) the event data is merged into the serializer hash for testing and [here](https://github.com/spree/spree/pull/11504/files#diff-d9f9039e1e67067424870ce8f4c2a70723bd047ac436880544b7755a714f6986R53) it's done right before making the webhook request.

File: api/db/migrate/20211118162826_change_spree_webhooks_events_null_defaults.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class ChangeSpreeWebhooksEventsNullDefaults < ActiveRecord::Migration[5.2]
+  def change
+    change_column_null :spree_webhooks_events, :subscriber_id, true
+    change_column_null :spree_webhooks_events, :url, true
```
Comment (damianlegawiec) [inline api/db/migrate/20211118162826_change_spree_webhooks_events_null_defaults.rb:4]: @sebastian-palma I think we don't need this change anymore?

File: api/spec/services/spree/webhooks/subscribers/handle_request_spec.rb (hunk 4 above)
Comment (damianlegawiec) [reply api/spec/services/spree/webhooks/subscribers/handle_request_spec.rb:60]: @sebastian-palma we need something more reliable than mocks here - we need a full test:

1. Order is placed
2. Job is created
3. Job is executed and we chang the payload being sent

So full E2E

File: api/app/services/spree/webhooks/subscribers/handle_request.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@ module Webhooks
     module Subscribers
       class HandleRequest
         def initialize(body:, event_name:, subscriber:)
-          @body = body
+          @body = JSON.parse(body)
```
Comment (damianlegawiec) [inline api/app/services/spree/webhooks/subscribers/handle_request.rb:8]: @sebastian-palma so this is the payload body right? not the response?

File: api/app/services/spree/webhooks/subscribers/queue_requests.rb [hunk 7]
```diff
@@ -6,7 +6,7 @@ class QueueRequests
 
         def call(body:, event:)
           Spree::Webhooks::Subscriber.active.with_urls_for(event).each do |subscriber|
-            Spree::Webhooks::Subscribers::MakeRequestJob.perform_later(body, event, subscriber)
+            Spree::Webhooks::Subscribers::MakeRequestJob.perform_later(body.to_json, event, subscriber)
```
Comment (damianlegawiec) [inline api/app/services/spree/webhooks/subscribers/queue_requests.rb:9]: @sebastian-palma I wonder, it's not recommended to push a lot of data as queue parameters, I wonder if this won't be a problem with bigger payloads, eg. if we would decide to by default include all associations for the record? To sum it up - are there any limits in the parameter size?

File: api/app/services/spree/webhooks/subscribers/handle_request.rb (hunk 6 above)
Comment (sebastian-palma) [reply api/app/services/spree/webhooks/subscribers/handle_request.rb:8]: The only purpose of this change is to avoid `ActiveJob::SerializationError`s as it raises an unsupported type error for the specific case of the order serializer (ActiveJob::SerializationError: Unsupported argument type: Spree::Money). We could convert those attributes to string but that would work only for OrderSerializer, if any other serializer has an unsupported type that can not automatically be serialized it'll result in the same error @damianlegawiec.

File: api/app/services/spree/webhooks/subscribers/queue_requests.rb (hunk 7 above)
Comment (sebastian-palma) [reply api/app/services/spree/webhooks/subscribers/queue_requests.rb:9]: I can't think of a limit size specifically @damianlegawiec, but we could modify this to avoid passing the body through the job arguments and instead to create the body in the job logic (HandleRequest) which is executed, so we could better follow the [best practices](https://github.com/mperham/sidekiq/wiki/Best-Practices#1-make-your-job-parameters-small-and-simple) for this.

Comment (damianlegawiec) [reply api/app/services/spree/webhooks/subscribers/queue_requests.rb:9]: @sebastian-palma when we create the body in the job we're risking we're sending a newer state of that resource so probably sending the payload as it is, is better

File: api/app/serializers/concerns/spree/api/v2/resource_serializer_concern.rb [hunk 8]
```diff
@@ -12,9 +12,20 @@ def self.included(base)
           # include standard attributes
           base.attributes(*model_klazz.json_api_columns)
           # include money attributes
-          base.attributes(*model_klazz.new.methods.find_all do |m|
```
Comment (damianlegawiec) [inline api/app/serializers/concerns/spree/api/v2/resource_serializer_concern.rb:15]: @sebastian-palma let's make a separate PR with this change

Comment (sebastian-palma) [reply api/app/serializers/concerns/spree/api/v2/resource_serializer_concern.rb:15]: sure @damianlegawiec, [here](https://github.com/spree/spree/pull/11523) is, I'm gonna cherry-pick those commits

**Review by damianlegawiec — APPROVED** [review]
